File: 0940-distinct-subsequences-ii/0940-distinct-subsequences-ii.java
class Solution {
    public int distinctSubseqII(String s) {
        int n = s.length();

        int dp = 0;

        int[] freq = new int[26];

        int MOD = 1000000007;

        for(char ch : s.toCharArray()){
            int new_dp = (2 * dp + 1 - freq[ch - 'a']) % MOD;

            if(new_dp < 0)
                new_dp += MOD;
                
            freq[ch - 'a'] = dp + 1;
            dp = new_dp;
        }

        return dp;
    }
}

File: 0940-distinct-subsequences-ii/0940-distinct-subsequences-ii.py
class Solution:
    def distinctSubseqII(self, s: str) -> int:
        dp = 0
        freq = [0] * 26

        MOD = 10**9 + 7

        for ch in s :

            new_dp = (2 * dp + 1 - freq[ord(ch) - ord('a')]) % MOD

            if new_dp < 0:
                new_dp += MOD

            freq[ord(ch) - ord('a')] = dp + 1
            dp = new_dp

        return dp